hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAssociateToAccountHandler.java

```diff
@@ -42,22 +67,223 @@ public TokenAssociateToAccountHandler() {
+            newTokenRels.add(lastOfNewTokenRels
+                    .copyBuilder()
+                    .nextToken(headTokenAsNonHeadTokenRel.tokenNumber())
+                    .build());
+            tokenRelStore.put(headTokenAsNonHeadTokenRel);
+        }
+
+        // Now replace the account's old head token number with the new head token number. This is
+        // how we link the new tokenRels to the account
+        final var firstOfNewTokenRels = newTokenRels.get(0);
+        final var updatedAcct = account.copyBuilder()
+                // replace the head token number with the first token number of the new tokenRels
+                .headTokenNumber(firstOfNewTokenRels.tokenNumber())
+                // and also update the account's total number of token associations
+                .numberAssociations(account.numberAssociations() + newTokenRels.size())
+                .build();
+
+        // Save the results
+        accountStore.put(updatedAcct);
+        newTokenRels.forEach(tokenRelStore::put);
+    }
+
+    /**
+     * Performs checks independent of state or context
+     */
+    private void pureChecks(@NonNull final TokenAssociateTransactionBody op) throws PreCheckException {
+        if (!op.hasAccount()) {
+            throw new PreCheckException(ResponseCodeEnum.INVALID_ACCOUNT_ID);
+        }
+
+        if (TokenListChecks.repeatsItself(op.tokensOrThrow())) {
+            throw new PreCheckException(TOKEN_ID_REPEATED_IN_TOKEN_LIST);
+        }
+    }
+
+    /**
+     * Performs checks that require state and context
+     */
+    @NonNull
+    private Validated validateSemantics(
+            @NonNull final List<TokenID> tokenIds,
+            @NonNull final AccountID accountId,
+            @NonNull final Configuration config,
+            @NonNull final WritableAccountStore accountStore,
+            @NonNull final ReadableTokenStore tokenStore,
+            @NonNull final WritableTokenRelationStore tokenRelStore) {
+        requireNonNull(config);
+
+        // Check that the system hasn't reached its limit of token associations
+        var maxTokenRels = config.getValue("maxNumTokenRels", Long.class);
+        validateTrue(
+                isTotalNumTokenRelsWithinMax(tokenIds.size(), tokenRelStore, maxTokenRels),
+                MAX_ENTITIES_IN_PRICE_REGIME_HAVE_BEEN_CREATED);
+
+        // Check that the account exists
+        final var account = accountStore.get(accountId).orElse(null);
+        validateTrue(account != null, INVALID_ACCOUNT_ID);
+
+        // Check that the given tokens exist and are usable
+        final var tokens = new ArrayList<Token>();
+        for (final TokenID tokenId : tokenIds) {
+            final var token = tokenStore.getToken(tokenId);
+            validateTrue(token != null, INVALID_TOKEN_ID);
+            validateFalse(token.deleted(), TOKEN_WAS_DELETED);
+            validateFalse(token.paused(), TOKEN_IS_PAUSED);
```

Comment:
We can add a method to tokenStore which is `ResponsecodeEnum getTokenIfUsable()`. Since this logic can be used in multiple places.

Reply:
Done. Added in `ContextualRetriever`